### PR TITLE
Homework 2: KNN

### PR DESCRIPTION
**Pull request by iraedeus**

### CODE REVIEW

File: ml_spbu/homework_2/KDTree/kd_max_heap.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+        heap = self.heap
+        fixed_x = self._fixed_x
+        metric = self._metric
+
+        dist_x_fixed_x = metric(x, fixed_x)
+
+        if not heap:
+            heap.append((x, dist_x_fixed_x))
+            return None
```
Comment (maybenotilya) [inline ml_spbu/homework_2/KDTree/kd_max_heap.py:107]: Cringe
```suggestion
            return
```

File: ml_spbu/homework_2/KDTree/kd_max_heap.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+        Args:
+            x (Point): The point to be added to the heap.
+        """
+
+        heap = self.heap
+        fixed_x = self._fixed_x
+        metric = self._metric
+
+        dist_x_fixed_x = metric(x, fixed_x)
```
Comment (maybenotilya) [inline ml_spbu/homework_2/KDTree/kd_max_heap.py:103]: Очень страшный нейминг

File: ml_spbu/homework_1/__init__.py
Comment (maybenotilya) [inline ml_spbu/homework_1/__init__.py:1]: Этого тут не должон быть

File: ml_spbu/homework_2/KDTree/kd_tree.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            X (list[Point]): List of data points to build the KD-Tree from.
+            leaf_size (int): The maximum number of points a leaf node can contain.
+            metric (CallableMetric): The distance metric function to be used for
+                nearest neighbor searches.
+        """
+
+        self._root: KDNode = KDNode(X, leaf_size)
+        self._metric: CallableMetric = metric
+
+    def knn_helper(self, x, k, current_node, k_nearest: KDMaxHeap):
```
Comment (maybenotilya) [inline ml_spbu/homework_2/KDTree/kd_tree.py:45]:
```suggestion
    def _knn_helper(self, x, k, current_node, k_nearest: KDMaxHeap):
```

File: ml_spbu/homework_2/KNNClassifier/knn_classifier.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    This class implements a K-Nearest Neighbors classifier that utilizes a KD-Tree
+    for efficient neighbor searching. It can be configured with different distance
+    metrics and leaf sizes for the KD-Tree.
+
+    Attributes:
+        _k (int): The number of neighbors to consider.
+        _leaf_size (int): The leaf size for the KD-Tree.
+        _metric (CallableMetric): The distance metric function.
+        _kd_tree (KDTree): The KD-Tree fitted on the training data.
+        _labels (dict[Point, int]): A dictionary mapping training points to their labels.
```
Comment (maybenotilya) [inline ml_spbu/homework_2/KNNClassifier/knn_classifier.py:25]: Ну во первых мы храним вдвое больше точек чем надо, во вторых кто запрещает иметь две одинаковые точки с разными лейблами

File: ml_spbu/homework_2/KNNClassifier/knn_classifier.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+        if not hasattr(self, "_kd_tree"):
+            raise NotFittedError()
+
+        output = []
+        results = self._kd_tree.query(X, self._k)
+
+        for result in results:
+            negatives = sum(self._labels[neighbor] == 0 for neighbor in result)
+            positives = sum(self._labels[neighbor] == 1 for neighbor in result)
```
Comment (maybenotilya) [inline ml_spbu/homework_2/KNNClassifier/knn_classifier.py:85]: Лейблов может быть больше двух

File: ml_spbu/homework_2/KNNClassifier/knn_classifier.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+        Returns:
+            list[int]: List of predicted class labels (0 or 1) for each input point.
+        """
+
+        output = []
+        labels = self.predict_proba(X)
+
+        for proba in labels:
+            output.append(int(proba[1] > proba[0]))
```
Comment (maybenotilya) [inline ml_spbu/homework_2/KNNClassifier/knn_classifier.py:108]: Тоже про больше двух лейблов

File: ml_spbu/homework_2/metrics/distance/minkowski.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+This module provides a function to compute the Minkowski distance, a generalized
+metric in normed vector space which is a generalization of both the Euclidean
+distance and the Manhattan distance.
+"""
+
+from ml_spbu.homework_2.annotations import Point
+
+
+def minkowski_distance(p1: Point, p2: Point, p: float) -> float:
```
Comment (maybenotilya) [inline ml_spbu/homework_2/metrics/distance/minkowski.py:11]: Бтв Евклидово расстояние и Манхеттенская метрика это частные случаи метрики Минковского, можно было сделать красиво и элегантно

File: ml_spbu/homework_2/normalizers/__init__.py
Comment (maybenotilya) [inline ml_spbu/homework_2/normalizers/__init__.py:1]: Почему `normalizers` а не `scalers`?

File: ml_spbu/homework_2/normalizers/abstract_scaler.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+        Args:
+            X (list[Point]): The input data to be transformed.
+
+        Returns:
+            list[Point]: The transformed data.
+        """
+
+    @abstractmethod
+    def fit_transform(self, X: list[Point]) -> list[Point]:
```
Comment (maybenotilya) [inline ml_spbu/homework_2/normalizers/abstract_scaler.py:51]:
```
self.fit(...)
return self.transform(...)
```

Comment (maybenotilya) [reply ml_spbu/homework_2/normalizers/abstract_scaler.py:51]: Оно абсолютно общее для всех

File: ml_spbu/homework_2/normalizers/minmax_scaler.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        X_scaled = []
+        n = self._features_number
+
+        data_min = self._data_min
+        data_max = self._data_max
+
+        for x in X:
+            x_scaled = []
+            for i in range(n):
+                if data_max[i] == data_min[i]:
```
Comment (maybenotilya) [inline ml_spbu/homework_2/normalizers/minmax_scaler.py:101]: nit: Для читабельности в целом это можно было и выше обработать

File: ml_spbu/homework_2/normalizers/standard_scaler.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        self._features_number = len(X[0])
+        n = self._features_number
+        m = len(X)
+
+        means = []
+        deviations = []
+
+        for i in range(n):
+            mean_i = sum(x[i] for x in X) / m
+            deviation_i = sqrt(sum((x[i] - mean_i) ** 2 for x in X) / m)
```
Comment (maybenotilya) [inline ml_spbu/homework_2/normalizers/standard_scaler.py:66]: Ну вообще можно было и из `statistics` взять

File: ml_spbu/homework_2/train_test_split.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+(features and labels) into separate training and testing sets.
+"""
+
+import random
+
+from ml_spbu.homework_2.annotations import Point
+
+
+def train_test_split(
+    X: list[Point], y: list[int], test_size: float = 0.2, random_state: int = 42
```
Comment (maybenotilya) [inline ml_spbu/homework_2/train_test_split.py:13]:
```suggestion
    X: list[Point], y: list[int], test_size: float = 0.2, random_state: int = 52
```

Comment (maybenotilya) [reply ml_spbu/homework_2/train_test_split.py:13]: Бтв кстати треин тест сплит из `sklearn` принимает любое число массивов данных

File: ml_spbu/homework_2/train_test_split.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+        >>> y_test
+        [0, 1]
+    """
+
+    if not (0 <= test_size <= 1):
+        raise ValueError("Test size must be from 0 to 1.")
+
+    X_labels = [(x, label) for x, label in zip(X, y)]
+    random.seed(random_state)
+    random.shuffle(X_labels)
```
Comment (maybenotilya) [inline ml_spbu/homework_2/train_test_split.py:56]: Можно же также просто индексы зашафлить и избежать лишних выделений памяти

File: ml_spbu/homework_2/notebooks/cancer.ipynb [hunk 13]
```diff
@@ -0,0 +1,718 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "0f6d5226-8448-497f-a143-31c822aceb55",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, 0.2, 42)"
```
Comment (maybenotilya) [inline ml_spbu/homework_2/notebooks/cancer.ipynb:563]:
```suggestion
    "X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, 0.2, 52)"
```

File: ml_spbu/homework_2/notebooks/cancer.ipynb [hunk 14]
```diff
@@ -0,0 +1,718 @@
...
+    "    knn = KNNClassifier(k, 5, euclidean_distance)\n",
+    "    knn.fit(X_train, y_train)\n",
+    "    y_pred = knn.predict(X_test)\n",
+    "\n",
+    "    accuracy_scores.append(accuracy_score(y_pred, y_test))\n",
+    "    f1_scores.append(f1_score(y_pred, y_test))\n",
+    "\n",
+    "plt.plot(range(1, 20), accuracy_scores)\n",
+    "plt.plot(range(1, 20), f1_scores)\n",
+    "plt.show()"
```
Comment (maybenotilya) [inline ml_spbu/homework_2/notebooks/cancer.ipynb:597]: Оси бы подписать

File: ml_spbu/homework_2/normalizers/minmax_scaler.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+    This transformation is often used as an alternative to zero mean,
+    unit variance scaling.
+
+    Attributes:
+        _data_min (list[float] | None): Minimum value for each feature in the training set.
+        _data_max (list[float] | None): Maximum value for each feature in the training set.
+        _features_number (int | None): Number of features in the input data.
+    """
+
+    def __init__(self):
```
Comment (maybenotilya) [inline ml_spbu/homework_2/normalizers/minmax_scaler.py:32]: Ну кстати оригинальный MinMaxScaler вроде приводит к диапозону [min, max], но так тоже сойдет

**Review by maybenotilya — COMMENTED** [review]
Основаная претензия к количеству классов в твоем классификаторе и лишним оверхедам по памяти, а так очень даже неплохо, мне нрав
